ldsocache.go

```
@@ -108,6 +139,193 @@ func (shdr *LDSOCacheExtensionSectionHeader) describe() {
 }
 
+func parse_lib_filename(realname string) (string, string, error) {
+	var name string
+	var ver string
+	// ldconfig(8) says it "will look only at files that are named lib*.so*
+	// (for regular shared objects) or ld-*.so* (for the dynamic loader itself).
+	// Other files will be ignored.
+	if !strings.HasPrefix(realname, "lib") && !strings.HasPrefix(realname, "ld-") {
+		return "", "", fmt.Errorf("filename does not start with 'lib' or 'ld-': %s", realname)
+	}
+	if !strings.Contains(realname, ".so") {
+		return "", "", fmt.Errorf("filename does not contain '.so': %s", realname)
+	}
+	if strings.HasSuffix(realname, ".so") {
+		name = strings.TrimSuffix(realname, ".so")
+		ver = ""
+	} else {
+		pieces := strings.Split(realname, ".so.")
+		if len(pieces) < 2 {
+			return "", "", fmt.Errorf("invalid library name: %s", realname)
+		}
+		name = strings.Join(pieces[:len(pieces)-1], ".so.")
+		ver = pieces[len(pieces)-1]
+	}
+
+	fmt.Printf("RETURNING %s for %s\n", name, realname)
+	return name, ver, nil
+}
+
+func AddLDSOCacheEntriesForDir(fsys fs.FS, libdir string, entries *[]LDSOCacheEntry) (error) {
+	// fs.FS wants all file paths to be relative
+	libdir = strings.TrimLeft(libdir, "/")
+	dirents, err := fs.ReadDir(fsys, libdir)
+	if err != nil {
+		// It is OK for a directory to not exist
+		return nil
+	}
+
+	for _, dirent := range dirents {
+		realname := dirent.Name()
+		fullpath := filepath.Join(libdir, realname)
+		mode := dirent.Type()
+		is_link := (mode & fs.ModeSymlink != 0)
+
+		if is_link {
+			// Stat follows symlinks
+			info, err := fs.Stat(fsys, fullpath)
+			if err != nil {
+				fmt.Printf("Warning: Could not stat %s\n", fullpath)
+				continue
+			}
+			if !info.Mode().IsRegular() {
+				fmt.Printf("DEBUG: Skipping %s, not a link to a regular file\n", fullpath)
+				continue
+			}
+		}
+
+		if !(mode.IsRegular() || is_link) {
+			continue
+		}
+		libf, err := fsys.Open(fullpath)
+		if err != nil {
+			continue
```

Is it okay to ignore this error?

---

I checked to see what `ldconfig` does by `chmod 000`'ing a lib:

```
$ ldconfig -C /tmp/cache
/sbin/ldconfig.real: Input file /lib/x86_64-linux-gnu/libpagemaker-0.0.so.0.0.4 not found.
$ echo $?
0
```

So I've added a warning here too.